runner/master_log.py

```diff
@@ -0,0 +1,172 @@
+from dataclasses import dataclass, field
+from datetime import datetime
+from typing import Optional, Dict, List
+
+from utils.constants import EventType
+
+
+@dataclass
+class LogEvent:
+    type: EventType
+    timestamp: datetime
+    test_id: Optional[str] = None
+    data: Optional[Dict[str, str]] = field(default_factory=dict)
+
+
+class MasterLog:
+
+    def __init__(self):
+        self.log: List[LogEvent] = []
+
+    def add_send_message(self, message: str, timestamp: datetime, test_id: Optional[str] = None, is_question: Optional[bool] = False):
+        if test_id:
+            event = LogEvent(EventType.SEND_MESSAGE, timestamp, test_id,  {"message": message, "is_question": is_question})
+        else:
+            event = LogEvent(EventType.SEND_FILL, timestamp, data={"message": message, "is_question": False})
+
+        self.add_event(event)
+
+    def add_response_message(self, message: str, timestamp: datetime, test_id: Optional[str] = None, is_question: Optional[bool] = False):
+        if test_id:
+            event = LogEvent(EventType.RESPONSE_MESSAGE, timestamp, test_id,  {"message": message, "is_question": is_question})
+        else:
+            event = LogEvent(EventType.RESPONSE_FILL, timestamp, data={"message": message, "is_question": False})
+
+        self.add_event(event)
+
+    def add_wait_event(self, test_id: str, timestamp: datetime, tokens=0, time=0):
+        event = LogEvent(EventType.WAIT, timestamp=timestamp, test_id=test_id, data={"tokens":tokens, "time":time})
+        self.add_event(event)
+
+    def begin_test(self, test_id, timestamp):
+        event = LogEvent(EventType.BEGIN, timestamp=timestamp, test_id=test_id)
+        self.add_event(event)
+
+    def end_test(self, test_id: str, timestamp: datetime):
+        event = LogEvent(EventType.END, timestamp=timestamp, test_id=test_id)
+        self.add_event(event)
+
+    def add_event(self, event: LogEvent):
+        self.log.append(event)
+
+    def human_readable_full_log(self, test_id: str, message: str) -> List[str]:
+        # Collate all the messages from the index point
+        messages = []
+        index = self.find_message(test_id, message)
+
+        for event in self.log[index:]:
+            if event.type in [EventType.SEND_MESSAGE, EventType.RESPONSE_MESSAGE, EventType.SEND_FILL, EventType.RESPONSE_FILL]:
+                sender = "Test" if event.type == EventType.SEND_MESSAGE else "Agent"
```

Comment:
This will put "Agent" when `event.type == EventType.SEND_FILL`. I was thinking that maybe two types (send and response) could be enough, being `fill` a flag.

Reply:
The rationale for FILL was because the filler was not tied to a test just to try and keep it cleaner. Good point on the agent part though.